### DISCUSSION TITLE
Unload a certificate loaded through the vici interface

### DISCUSSION THREAD

**Discussion by bhatnitish** [strongswan/strongswan#1935] [Q&A]
Hi,

I'm loading client and CA certificates with the connection through the load-conn vici command (providing the cert information in local/remote auth sections). I then load the keys using the load-key command. Is there any way I can clean up these certs one at a time after I unload the connection? In my use case, I have multiple connections being loaded/unloaded and the certificates would accumulate over time. The clear-creds command will not work for me as it'll just clear creds for existing connections too.

**Comment by tobiasbrunner** [comment]
> I'm loading client and CA certificates with the connection through the load-conn vici command (providing the cert information in local/remote auth sections).

While loading CA certificates that way also loads them to the trust store, this primarily sets a CA constraint for the peer certificate. If you want to be able to unload individual CA certificates, I'd recommend loading them via `load-authority` instead (such certificates also won't be affected by calls to `clear-creds`).

> Is there any way I can clean up these certs one at a time after I unload the connection?

Client certificates loaded via `load-conn` are unloaded when the connection is unloaded. They might be cached in the in-memory cache, but you can clear that via `flush-certs` command if it's a concern for you. CA certificates loaded with connections won't be unloaded automatically as they might be shared by multiple connections, that requires an explicit `clear-creds` call (as mentioned above, CA certificates loaded via `load-authority` are not affected by that).

>  In my use case, I have multiple connections being loaded/unloaded and the certificates would accumulate over time. 

Why is that a problem? How many CAs will there be?

**Reply by bhatnitish** [reply]
Hi Tobias, thanks for the reply. I have now started using load/unload-authority for CA certificates. For the client certificates loaded through the local/remote auth sections in load-conn, I still see them when I run --list-certs. This is after I unload the conn and flush the certs. I am running strongswan 5.9.7. Is there anything I'm doing wrong here? Here are the strongswan logs (these are dummy keys):

`
Oct  6 09:09:13.445 11[CFG2] vici client 1 requests: load-conn

Oct  6 09:09:13.445 11[CFG2]  conn ipsecPolicy2:

Oct  6 09:09:13.446 11[CFG1]   id not specified, defaulting to cert subject 'C=CH, O=strongSwan, CN=device1'
Oct  6 09:09:13.446 11[CFG2]   child IPSec-default/default/ipsecPdtPolicy2:
Oct  6 09:09:13.446 11[CFG2]    rekey_time = 3600
Oct  6 09:09:13.446 11[CFG2]    life_time = 3960
Oct  6 09:09:13.446 11[CFG2]    rand_time = 360
Oct  6 09:09:13.446 11[CFG2]    rekey_bytes = 0
Oct  6 09:09:13.446 11[CFG2]    life_bytes = 0
Oct  6 09:09:13.446 11[CFG2]    rand_bytes = 0
Oct  6 09:09:13.446 11[CFG2]    rekey_packets = 0
Oct  6 09:09:13.446 11[CFG2]    life_packets = 0
Oct  6 09:09:13.446 11[CFG2]    rand_packets = 0
Oct  6 09:09:13.446 11[CFG2]    updown = (null)
Oct  6 09:09:13.446 11[CFG2]    hostaccess = 0
Oct  6 09:09:13.446 11[CFG2]    ipcomp = 0
Oct  6 09:09:13.446 11[CFG2]    mode = TUNNEL
Oct  6 09:09:13.446 11[CFG2]    policies = 0
Oct  6 09:09:13.446 11[CFG2]    policies_fwd_out = 0
Oct  6 09:09:13.446 11[CFG2]    dpd_action = none
Oct  6 09:09:13.446 11[CFG2]    start_action = none
Oct  6 09:09:13.446 11[CFG2]    close_action = none
Oct  6 09:09:13.446 11[CFG2]    reqid = 0
Oct  6 09:09:13.446 11[CFG2]    tfc = 0
Oct  6 09:09:13.446 11[CFG2]    priority = 0
Oct  6 09:09:13.446 11[CFG2]    interface = (null)
Oct  6 09:09:13.446 11[CFG2]    if_id_in = 0
Oct  6 09:09:13.446 11[CFG2]    if_id_out = 0
Oct  6 09:09:13.446 11[CFG2]    mark_in = 0/0
Oct  6 09:09:13.446 11[CFG2]    mark_in_sa = 0
Oct  6 09:09:13.446 11[CFG2]    mark_out = 0/0
Oct  6 09:09:13.446 11[CFG2]    set_mark_in = 0/0
Oct  6 09:09:13.446 11[CFG2]    set_mark_out = 0/0
Oct  6 09:09:13.446 11[CFG2]    label = (null)
Oct  6 09:09:13.446 11[CFG2]    label_mode = system
Oct  6 09:09:13.446 11[CFG2]    inactivity = 0
Oct  6 09:09:13.446 11[CFG2]    proposals = ESP:AES_GCM_16_128/ECP_384/EXT_SEQ/NO_EXT_SEQ
Oct  6 09:09:13.446 11[CFG2]    local_ts = 0.0.0.0/0
Oct  6 09:09:13.446 11[CFG2]    remote_ts = 0.0.0.0/0
Oct  6 09:09:13.446 11[CFG2]    hw_offload = no
Oct  6 09:09:13.446 11[CFG2]    sha256_96 = 0
Oct  6 09:09:13.446 11[CFG2]    copy_df = 1
Oct  6 09:09:13.446 11[CFG2]    copy_ecn = 1
Oct  6 09:09:13.446 11[CFG2]    copy_dscp = out
Oct  6 09:09:13.446 11[CFG2]   version = 2
Oct  6 09:09:13.446 11[CFG2]   local_addrs = 169.254.13.116
Oct  6 09:09:13.446 11[CFG2]   remote_addrs = 169.254.13.6
Oct  6 09:09:13.446 11[CFG2]   local_port = 500
Oct  6 09:09:13.446 11[CFG2]   remote_port = 500
Oct  6 09:09:13.446 11[CFG2]   send_certreq = 1
Oct  6 09:09:13.446 11[CFG2]   send_cert = CERT_SEND_IF_ASKED
Oct  6 09:09:13.446 11[CFG2]   ppk_id = (null)
Oct  6 09:09:13.446 11[CFG2]   ppk_required = 0
Oct  6 09:09:13.446 11[CFG2]   mobike = 0
Oct  6 09:09:13.446 11[CFG2]   aggressive = 0
Oct  6 09:09:13.446 11[CFG2]   dscp = 0x00
Oct  6 09:09:13.446 11[CFG2]   encap = 0
Oct  6 09:09:13.446 11[CFG2]   dpd_delay = 60
Oct  6 09:09:13.446 11[CFG2]   dpd_timeout = 0
Oct  6 09:09:13.446 11[CFG2]   fragmentation = 0
Oct  6 09:09:13.446 11[CFG2]   childless = 0
Oct  6 09:09:13.446 11[CFG2]   unique = UNIQUE_KEEP
Oct  6 09:09:13.446 11[CFG2]   keyingtries = 1
Oct  6 09:09:13.446 11[CFG2]   reauth_time = 86400
Oct  6 09:09:13.446 11[CFG2]   rekey_time = 14400
Oct  6 09:09:13.446 11[CFG2]   over_time = 8640
Oct  6 09:09:13.446 11[CFG2]   rand_time = 7200
Oct  6 09:09:13.446 11[CFG2]   proposals = IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_384
Oct  6 09:09:13.446 11[CFG2]   if_id_in = 0
Oct  6 09:09:13.446 11[CFG2]   if_id_out = 0
Oct  6 09:09:13.446 11[CFG2]   local:
Oct  6 09:09:13.446 11[CFG2]    id = C=CH, O=strongSwan, CN=device1
Oct  6 09:09:13.446 11[CFG2]    class = public key
Oct  6 09:09:13.446 11[CFG2]    cert = C=CH, O=strongSwan, CN=device1  **>>> This is the client cert**
Oct  6 09:09:13.446 11[CFG2]   remote:
Oct  6 09:09:13.446 11[CFG2]    class = public key
Oct  6 09:09:13.446 11[CFG1] added vici connection: ipsecPolicy2
Oct  6 09:09:13.446 17[CFG2] vici client 1 requests: load-key
Oct  6 09:09:13.454 17[CFG1] loaded RSA private key  **>>> Load private key for client certificate**
Oct  6 09:09:13.454 19[CFG2] vici client 1 requests: load-authority  **>>> Load CA certificate**
Oct  6 09:09:13.454 19[CFG2]  authority 4666912421ede4e446a29e56283bceb76c4d4e446a29e521b286a83bceb74666:
Oct  6 09:09:13.454 19[CFG2]   cacert = C=CH, O=strongSwan, CN=Root CA
Oct  6 09:09:24.308 10[CFG2] vici client 12 connected
Oct  6 09:09:24.308 12[CFG2] vici client 12 registered for: list-conn
Oct  6 09:09:24.308 15[CFG2] vici client 12 requests: list-conns
Oct  6 09:09:24.310 19[CFG2] vici client 12 disconnected
Oct  6 09:09:35.868 05[CFG2] vici client 13 connected
Oct  6 09:09:35.868 09[CFG2] vici client 13 registered for: list-cert
Oct  6 09:09:35.869 11[CFG2] vici client 13 requests: list-certs
Oct  6 09:09:35.871 15[CFG2] vici client 13 disconnected
Oct  6 09:10:22.730 08[CFG2] vici client 1 registered for: list-conn
Oct  6 09:10:22.730 05[CFG2] vici client 1 requests: list-conns
Oct  6 09:10:22.730 10[CFG2] vici client 1 unregistered for: list-conn
Oct  6 09:10:22.731 14[CFG2] vici client 1 requests: unload-conn  **>>> Unload the connection, keys here**
Oct  6 09:10:22.731 08[CFG2] vici client 1 requests: unload-key
Oct  6 09:10:22.731 08[CFG1] unloaded private key with id bd3d5414aa7802b770bd9aa110db3dd525779e49
Oct  6 09:10:22.731 18[CFG2] vici client 1 requests: unload-authority  **>>> Unload-authority**
Oct  6 09:10:22.731 10[CFG2] vici client 1 requests: flush-certs  **>>> Flush-certs**
Oct  6 09:10:27.712 04[CFG2] vici client 14 connected
Oct  6 09:10:27.713 09[CFG2] vici client 14 registered for: list-cert
Oct  6 09:10:27.713 11[CFG2] vici client 14 requests: list-certs
Oct  6 09:10:27.714 17[CFG2] vici client 14 disconnected
`

**Reply by tobiasbrunner** [reply]
You're right, the connection-specific certificates are also stored in a global credential set (to avoid storing duplicate certificates when multiple connections share the same ones). Certificates there can only be removed via `clear-creds`.
